### ISSUE TITLE
Migration requires allowAdminChanges

### ISSUE THREAD

**Issue by engram-design** [barrelstrength/craft-sprout-forms#312]
We just hit a migration error when trying to push updates to a staging server, where `allowAdminChanges` was false. This crashed the site, and the only way to fix this was to set `allowAdminChanges` to true and run the migration again.

```
output > create table {{%sproutforms_integrations}} ...
output done (time: 0.003s)
output > create index sproutforms_integrations_formId_fk on {{%sproutforms_integrations}} (formId) ...
output done (time: 0.003s)
output > add foreign key sproutforms_integrations_formId_fk: {{%sproutforms_integrations}} (formId) references {{%sproutforms_forms}} (id) ...
output done (time: 0.013s)
output Exception: Changes to the project config are not possible while in read-only mode. (/home/forge/site/vendor/craftcms/cms/src/services/ProjectConfig.php:354)
output #0 /home/forge/site/vendor/barrelstrength/sprout-forms/src/migrations/m190410_000000_add_payload_forwarding_to_integration.php(77): craft\services\ProjectConfig->set('plugins.sprout-...', Array)
output #1 /home/forge/site/vendor/craftcms/cms/src/db/Migration.php(56): barrelstrength\sproutforms\migrations\m190410_000000_add_payload_forwarding_to_integration->safeUp()
output #2 /home/forge/site/vendor/craftcms/cms/src/db/MigrationManager.php(243): craft\db\Migration->up(true)
output #3 /home/forge/site/vendor/craftcms/cms/src/db/MigrationManager.php(163): craft\db\MigrationManager->migrateUp(Object(barrelstrength\sproutforms\migrations\m190410_000000_add_payload_forwarding_to_integration))
output #4 /home/forge/site/vendor/craftcms/cms/src/services/Updates.php(225): craft\db\MigrationManager->up()
output #5 /home/forge/site/vendor/craftcms/cms/src/console/controllers/MigrateController.php(259): craft\services\Updates->runMigrations(Array)
output #6 [internal function]: craft\console\controllers\MigrateController->actionAll()
output #7 /home/forge/site/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
output #8 /home/forge/site/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
output #9 /home/forge/site/vendor/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('all', Array)
output #10 /home/forge/site/vendor/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('all', Array)
output #11 /home/forge/site/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('migrate/all', Array)
output #12 /home/forge/site/vendor/craftcms/cms/src/console/Application.php(93): yii\console\Application->runAction('migrate/all', Array)
output #13 /home/forge/site/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('migrate/all', Array)
output #14 /home/forge/site/vendor/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(craft\console\Request))
output #15 /home/forge/site/craft(22): yii\base\Application->run()
output #16 {main}
warning Exception 'craft\errors\MigrateException' with message 'An error occurred while migrating Sprout Forms.'
warning
warning in /home/forge/site/vendor/craftcms/cms/src/services/Updates.php:233
warning
warning Stack trace:
warning #0 /home/forge/site/vendor/craftcms/cms/src/console/controllers/MigrateController.php(259): craft\services\Updates->runMigrations(Array)
warning #1 [internal function]: craft\console\controllers\MigrateController->actionAll()
warning #2 /home/forge/site/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
warning #3 /home/forge/site/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
warning #4 /home/forge/site/vendor/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('all', Array)
warning #5 /home/forge/site/vendor/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('all', Array)
warning #6 /home/forge/site/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('migrate/all', Array)
warning #7 /home/forge/site/vendor/craftcms/cms/src/console/Application.php(93): yii\console\Application->runAction('migrate/all', Array)
warning #8 /home/forge/site/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('migrate/all', Array)
warning #9 /home/forge/site/vendor/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(craft\console\Request))
warning #10 /home/forge/site/craft(22): yii\base\Application->run()
warning #11 {main}
execute ./craft migrate/all
```

**Comment by BenParizek** [comment]
@engram-design I think I'm missing something here, but I'm not sure of the best way to solve this yet.

The migration that is being run is both adding new tables, which would need to be a migration that runs in all environments, and updating the Project Config, which, if enabled should only happen in the permitted environments.

Should all migrations we write be able to run with `allowAdminChanges` disabled? If so, where do we write a migration when we need to modify the Project Config settings that may only be permitted in a single environment?

I'll ask around on this and update here once we figure out the best path forward.

**Comment by BenParizek** [comment]
Related:
https://github.com/craftcms/cms/issues/3653
https://github.com/putyourlightson/craft-blitz/issues/114

No clear solution yet.

**Comment by BenParizek** [comment]
Here's the documentation on how to handle this:
https://docs.craftcms.com/v3/extend/project-config.html#project-config-migrations

We need to add a versionCompare statment and only run the Project Config migration if it hasn't already been run. The user will need to make sure this happens in the local environment workflow.

**Comment by BenParizek** [comment]
@engram-design This should be addressed in the latest release v3.4.0.

**Comment by kgi-cwells** [comment]
Hi, This issue popped up again when attempting to upgrade from v3.4.3 to v3.5.1, with a migration not properly checking the plugin schema, thereby forcing an exception when trying to migrate when `allowAdminChanges=false`.

I was able to fix it by manually editing the migration file `m191005_000000_update_form_settings.php` and correcting it per [https://docs.craftcms.com/v3/extend/project-config.html#project-config-migrations](https://docs.craftcms.com/v3/extend/project-config.html#project-config-migrations). I can send a PR if needed!

For reference, the exception I was getting:

```
$ ./craft migrate/all
Yii Migration Tool (based on Yii v2.0.21)

Exception: Changes to the project config are not possible while in read-only mode. (/mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/services/ProjectConfig.php:355)
#0 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/barrelstrength/sprout-forms/src/migrations/m191005_000000_update_form_settings.php(53): craft\services\ProjectConfig->set('plugins.sprout-...', Array)
#1 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/db/Migration.php(58): barrelstrength\sproutforms\migrations\m191005_000000_update_form_settings->safeUp()
#2 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/db/MigrationManager.php(239): craft\db\Migration->up(true)
#3 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/db/MigrationManager.php(159): craft\db\MigrationManager->migrateUp(Object(barrelstrength\sproutforms\migrations\m191005_000000_update_form_settings))
#4 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/services/Updates.php(225): craft\db\MigrationManager->up()
#5 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/console/controllers/MigrateController.php(265): craft\services\Updates->runMigrations(Array)
#6 [internal function]: craft\console\controllers\MigrateController->actionAll()
#7 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#8 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#9 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('all', Array)
#10 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('all', Array)
#11 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('migrate/all', Array)
#12 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/console/Application.php(93): yii\console\Application->runAction('migrate/all', Array)
#13 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('migrate/all', Array)
#14 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(craft\console\Request))
#15 /mnt/klipschweb-data-prod1/www/klipsch3-staging/craft(22): yii\base\Application->run()
#16 {main}
Exception 'craft\errors\MigrateException' with message 'An error occurred while migrating Sprout Forms.'

in /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/services/Updates.php:233

Stack trace:
#0 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/console/controllers/MigrateController.php(265): craft\services\Updates->runMigrations(Array)
#1 [internal function]: craft\console\controllers\MigrateController->actionAll()
#2 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#3 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams(Array)
#4 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/console/Controller.php(148): yii\base\Controller->runAction('all', Array)
#5 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/Module.php(528): yii\console\Controller->runAction('all', Array)
#6 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/console/Application.php(180): yii\base\Module->runAction('migrate/all', Array)
#7 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/craftcms/cms/src/console/Application.php(93): yii\console\Application->runAction('migrate/all', Array)
#8 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/console/Application.php(147): craft\console\Application->runAction('migrate/all', Array)
#9 /mnt/klipschweb-data-prod1/www/klipsch3-staging/vendor/yiisoft/yii2/base/Application.php(386): yii\console\Application->handleRequest(Object(craft\console\Request))
#10 /mnt/klipschweb-data-prod1/www/klipsch3-staging/craft(22): yii\base\Application->run()
#11 {main}
```
